filestorage.go

```diff
@@ -149,125 +152,112 @@ func dataDir() string {
 	return filepath.Join(baseDir, "certmagic")
 }
 
-// TryLock attempts to get a lock for name, otherwise it returns
-// a Waiter value to wait until the other process is finished.
-func (fs FileStorage) TryLock(key string) (Waiter, error) {
-	fileStorageNameLocksMu.Lock()
-	defer fileStorageNameLocksMu.Unlock()
+// Lock obtains a lock named by the given key. It blocks
+// until the lock can be obtained or an error is returned.
+func (fs *FileStorage) Lock(key string) error {
+	// can't defer the unlock because we might have
+	// to Wait() for a while before returning, so we're
+	// careful to unlock at all the right places
+	fs.fileStorageNameLocksMu.Lock()
+
+	if fs.fileStorageNameLocks == nil {
+		fs.fileStorageNameLocks = make(map[string]*fileStorageWaiter)
+	}
 
 	// see if lock already exists within this process - allows
 	// for faster unlocking since we don't have to poll the disk
-	fw, ok := fileStorageNameLocks[key]
+	fw, ok := fs.fileStorageNameLocks[key]
 	if ok {
 		// lock already created within process, let caller wait on it
-		return fw, nil
+		fs.fileStorageNameLocksMu.Unlock()
+		fw.Wait()
+		return nil
 	}
 
 	// attempt to persist lock to disk by creating lock file
 
-	// parent dir must exist
 	lockDir := fs.lockDir()
-	if err := os.MkdirAll(lockDir, 0700); err != nil {
-		return nil, err
-	}
-
+	// since there isn't already a waiter for the lock, make one
 	fw = &fileStorageWaiter{
 		key:      key,
 		filename: filepath.Join(lockDir, StorageKeys.safe(key)+".lock"),
 		wg:       new(sync.WaitGroup),
 	}
+	fw.wg.Add(1)
+	fs.fileStorageNameLocks[key] = fw
+	fs.fileStorageNameLocksMu.Unlock()
 
-	var checkedStaleLock bool // sentinel value to avoid infinite goto-ing
-
-createLock:
-	// create the file in a special mode such that an
-	// error is returned if it already exists
-	lf, err := os.OpenFile(fw.filename, os.O_CREATE|os.O_EXCL, 0644)
-	if err != nil {
-		if os.IsExist(err) {
-			// another process has the lock
-
-			// check to see if the lock is stale, if we haven't already
-			if !checkedStaleLock {
-				checkedStaleLock = true
-				if fs.lockFileStale(fw.filename) {
-					log.Printf("[INFO][%s] Lock for '%s' is stale; removing then retrying: %s",
-						fs, key, fw.filename)
-					os.Remove(fw.filename)
-					goto createLock
-				}
+	for {
+		if createLockfile(fw.filename) == nil {
+			return nil
+		}
+		// we'll just assume the lockfile exists
+		info, err := os.Stat(fw.filename)
+		switch {
+		case err != nil:
+			// we assume the lockfile no longer exists; if we're now able to create it, great!
+			err := createLockfile(fw.filename)
+			if err != nil {
+				// we called wg.Add(1) above but didn't actually acquire the lock
+				fw.wg.Done()
 			}
-
-			// if lock is not stale, wait upon it
-			return fw, nil
+			return err
```

Comment:
One thing I just realized about this change here is that it could erroneously return an error.

For example, to get here we first tried to create the lock on disk which failed, presumably because it already existed; then we tried to stat it, which failed, so it was probably removed; then we tried to create it again, which failed, so... what does that mean? Maybe it was created again by another process, but here we just return an error, even though the correct thing to do would be to wait longer.

I'll push a commit soon which addresses this, I think. Hopefully for our final round of reviews! :+1: